evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java

```diff
@@ -29,45 +32,105 @@
+
+  /** The HISTORY_SERVE_WINDOW */
+  public static final long HISTORY_SERVE_WINDOW = 8192;
+
+  /** The BLOCKHASH_OLD_WINDOW */
+  public static final int BLOCKHASH_OLD_WINDOW = 256;
+
+  private static final int MAX_BLOCK_ARG_SIZE = 8;
+
+  private final BlockHashRetrievalStrategy blockHashRetrievalStrategy;
 
   /**
    * Instantiates a new Block hash operation.
    *
    * @param gasCalculator the gas calculator
    */
   public BlockHashOperation(final GasCalculator gasCalculator) {
+    this(gasCalculator, BLOCK_HASH_LOOKUP);
+  }
+
+  /**
+   * Instantiates a new Block hash operation.
+   *
+   * @param gasCalculator the gas calculator
+   * @param blockHashRetrievalStrategy whether read from state (EIP-2935)
+   */
+  public BlockHashOperation(
+      final GasCalculator gasCalculator,
+      final BlockHashRetrievalStrategy blockHashRetrievalStrategy) {
     super(0x40, "BLOCKHASH", 1, 1, gasCalculator, gasCalculator.getBlockHashOperationGasCost());
+    this.blockHashRetrievalStrategy = blockHashRetrievalStrategy;
   }
 
   @Override
-  public Operation.OperationResult executeFixedCostOperation(
-      final MessageFrame frame, final EVM evm) {
+  public OperationResult executeFixedCostOperation(final MessageFrame frame, final EVM evm) {
     final Bytes blockArg = frame.popStackItem().trimLeadingZeros();
 
-    // Short-circuit if value is unreasonably large
-    if (blockArg.size() > 8) {
+    if (blockArg.size() > MAX_BLOCK_ARG_SIZE) {
       frame.pushStackItem(UInt256.ZERO);
       return successResponse;
     }
 
     final long soughtBlock = blockArg.toLong();
-    final BlockValues blockValues = frame.getBlockValues();
-    final long currentBlockNumber = blockValues.getNumber();
-    final long mostRecentBlockNumber = currentBlockNumber - 1;
-
-    // If the current block is the genesis block or the sought block is
-    // not within the last 256 completed blocks, zero is returned.
-    if (currentBlockNumber == 0
-        || soughtBlock < (mostRecentBlockNumber - MAX_RELATIVE_BLOCK)
-        || soughtBlock > mostRecentBlockNumber) {
-      frame.pushStackItem(Bytes32.ZERO);
+    final long currentBlockNumber = frame.getBlockValues().getNumber();
+
+    if (!isBlockWithinServeWindow(soughtBlock, currentBlockNumber)) {
+      frame.pushStackItem(UInt256.ZERO);
     } else {
-      final Function<Long, Hash> blockHashLookup = frame.getBlockHashLookup();
-      final Hash blockHash = blockHashLookup.apply(soughtBlock);
-      frame.pushStackItem(blockHash);
+      frame.pushStackItem(getBlockHash(frame, soughtBlock));
     }
 
     return successResponse;
   }
+
+  private boolean isBlockWithinServeWindow(final long soughtBlock, final long currentBlockNumber) {
+
+    final long historyServeWindow =
+        isServingFromState() ? HISTORY_SERVE_WINDOW : BLOCKHASH_OLD_WINDOW;
+
+    return soughtBlock >= Math.max(currentBlockNumber - historyServeWindow, 0)
+        && soughtBlock < currentBlockNumber;
```

Comment:
The serve window should include the currentBlockNumber as we used to do that from what I can see of the previous code. So this should be changed from `soughtBlock < currentBlockNumber` to `soughtBlock <= currentBlockNumber` reflecting the negation of the previous condition `soughtBlock > mostRecentBlockNumber`.

Reply:
current block is supposed to be a zero hash return, per the yellowpaper.

Reply:
yep your right it is defined that way